File: solution/449/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:
    void change_int_to_string(int num, string &str_val){
        string temp;
        while(num){
            temp += num%10 + '0';
            num = num/10;
        }
        for(int i=temp.length()-1;i>=0;i--){
            str_val += temp[i];
        }
        str_val += '#';
    }

    void preorder_serialize(TreeNode* root, string &str){
        if(!root)return;
        string str_val;
        change_int_to_string(root->val, str_val);
        str += str_val;
        preorder_serialize(root->left, str);
        preorder_serialize(root->right, str);

    }

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string str;
        preorder_serialize(root, str);
        return str;
    }

    void BST_insert(TreeNode* node, TreeNode* insertnode){
        if(insertnode->val > node->val){
            if(node->right){
                BST_insert(node->right, insertnode);
            }else{
                node->right = insertnode;
            }
        }else{
            if(node->left){
                BST_insert(node->left, insertnode);
            }else{
                node->left = insertnode;
            }
        }
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if(data.length()==0)return NULL;
        vector<TreeNode*> nodelist;
        int temp=0;
        for(int i=0;i<data.length();i++){
            if(data[i]!='#')temp = temp*10 + (data[i] - '0');
            else{
                nodelist.push_back(new TreeNode(temp));
                temp = 0;
            }
        }
        for(int i=1;i<nodelist.size();i++){
            BST_insert(nodelist[0], nodelist[i]);
        }
        return nodelist[0];
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));

File: solution/449/solution.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:
    def preorder_serialize(self, node, allnum):
        if not node:return
        allnum.append(str(node.val))
        self.preorder_serialize(node.left, allnum)
        self.preorder_serialize(node.right, allnum)

    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        allnum = []
        self.preorder_serialize(root, allnum)
        return ("#").join(allnum)

    def BST_insert(self, node, insertnode):
        if insertnode.val < node.val:
            if node.left:
                self.BST_insert(node.left, insertnode)
            else:
                node.left = insertnode
        else:
            if node.right:
                self.BST_insert(node.right, insertnode)
            else:
                node.right = insertnode

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        if not data:return None
        datanums = data.split("#")
        begin = TreeNode(int(datanums[0]))
        for x in datanums[1:]:
            self.BST_insert(begin, TreeNode(int(x)))
        return begin


# Your Codec object will be instantiated and called as such:
a = TreeNode(5)
b = TreeNode(2)
c = TreeNode(13)
a.left = b
a.right = c
codec = Codec()
d = codec.deserialize(codec.serialize(a))
print d.val
print d.left.val
print d.right.val